ipq806x: add support for Buffalo WXR-2533DHP

musashino205 opened the pull request:
Buffalo WXR-2533DHP is a 2.4/5 GHz band 11ac router, based on Qualcomm
IPQ8064.

The U-Boot on WXR-2533DHP behaves slightly complicated according to
the conditions of os-images. See the notes in buffalo.sh for details.

specifications:

- Qualcomm IPQ8064 (384 - 1,400 MHz, 2C2T)
- 512 MB of RAM (DDR3)
- 256 MB of Flash (NAND)
- 4T4R 2.4/5 GHz Wlan (QCA9980)
- 5x 10/100/1000 Mbps Ethernet
- 10x LEDs, 8x keys (6x buttons, 2x slide-switches)
- 2x USB 3.0 Type-A
- 12VDC/4A AC Adapter
- UART through-hole on PCB
  - J3: Vcc, GND, TX, RX from USB port side
  - 115200n8

Flash instruction using initramfs image:

1. Prepare the TFTP server with initramfs image renamed to
"wxr2300dhp-initramfs.uImage" and IP address "192.168.11.10"
2. Push power button on WXR-2533DHP with holding the "AOSS" button
and turn on WXR-2533DHP
3. Flash "Wireless" LED and release finger from AOSS button,
WXR-2533DHP downloads renamed initramfs image from TFTP server and
boot with it
4. On the initramfs image, perform sysupgrade with squashfs-sysupgrade
image
5. Wait ~120 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
ipq806x: add support for Buffalo WXR-2533DHP

Buffalo WXR-2533DHP is a 2.4/5 GHz band 11ac router, based on Qualcomm
IPQ8064.

The U-Boot on WXR-2533DHP behaves slightly complicated according to
the conditions of os-images. See the notes in buffalo.sh for details.

specifications:

- Qualcomm IPQ8064 (384 - 1,400 MHz, 2C2T)
- 512 MB of RAM (DDR3)
- 256 MB of Flash (NAND)
- 4T4R 2.4/5 GHz Wlan (QCA9980)
- 5x 10/100/1000 Mbps Ethernet
- 10x LEDs, 8x keys (6x buttons, 2x slide-switches)
- 2x USB 3.0 Type-A
- 12VDC/4A AC Adapter
- UART through-hole on PCB
  - J3: Vcc, GND, TX, RX from USB port side
  - 115200n8

Flash instruction using initramfs image:

1. Prepare the TFTP server with initramfs image renamed to
"wxr2300dhp-initramfs.uImage" and IP address "192.168.11.10"
2. Push power button on WXR-2533DHP with holding the "AOSS" button
and turn on WXR-2533DHP
3. Flash "Wireless" LED and release finger from AOSS button,
WXR-2533DHP downloads renamed initramfs image from TFTP server and
boot with it
4. On the initramfs image, perform sysupgrade with squashfs-sysupgrade
image
5. Wait ~120 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

chunkeey commented:
Have you abandoned https://github.com/openwrt/openwrt/pull/1609 ?

musashino205 commented:
No, for me it is still going on...

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+	model = "Buffalo WXR-2533DHP";
+	compatible = "buffalo,wxr-2533dhp", "qcom,ipq8064";
+
+	memory@0 {
```
unit-address "0" does not match reg.

chunkeey commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -11,6 +11,11 @@ board=$(board_name)
 
 case "$board" in
+buffalo,wxr-2533dhp)
+	ucidef_set_led_usbport "usb" "USB" "${boardname}:green:usb" "usb1-port1" "usb2-port1" "usb3-port1" "usb4-port1"
```
You could try to see if you can declare the usbport trigger via DT. 
https://github.com/openwrt/openwrt/blob/0d23fd2ab29a66f5d03187db4fac3e396b4f3b62/target/linux/ath79/dts/qca9558_tplink_archer-c7.dtsi#L41

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+
+	aliases {
+		serial0 = &gsbi4_serial;
+		mdio-gpio0 = &mdio0;
```
Aliases are usually meant for user-accessible devices/ports/interfaces.

https://patchwork.kernel.org/patch/9133903/#19207021
"If those ports are physically organised and labelled the same, then
using aliases could make sense, to describe the well-defined physical
labels. If you've assigned the numbers artificially, or if the physical
organisation differs across boards, then aliases are not the right tool
for the job. "

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+	};
+
+	chosen {
+		linux,stdout-path = "serial0:115200n8";
```
linux,stdout-path is deprecated.
https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicenodes.rst#chosen-node

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+			pinctrl-0 = <&nand_pins>;
+			pinctrl-names = "default";
+
+			cs0 {
```
node-name lacks unit-address @0 (since reg = <0>)

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+			compatible = "virtual,mdio-gpio";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			gpios = <&qcom_pinmux 1 GPIO_ACTIVE_HIGH &qcom_pinmux 0 GPIO_ACTIVE_HIGH>;
```
please make that

gpios = <&qcom_pinmux 1 GPIO_ACTIVE_HIGH>,
		<&qcom_pinmux 0 GPIO_ACTIVE_HIGH>;

(convert to tabs

chunkeey commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -72,6 +72,20 @@ define Device/ZyXELImage
+	IMAGE_SIZE := 65536k
+	KERNEL_IN_UBI := 1
+	IMAGES := sysupgrade.bin
+	DEVICE_PACKAGES := ath10k-firmware-qca99x0-ct squashfs-tools-mksquashfs
```
In your flashing instructions you state that the user has to flash via the initramfs and then use sysupgrade.
Problem is the initramfs image is not enabled by default for the ipq806x targets since it does not have the ramdisk feature in the target's makefile. 

If this got fixed and FEATURES is extended . Then unfortunately, the initramfs images that the OpenWrt buildbots are creating for the official images (CONFIG_TARGET_ALL_PROFILES!) will not have the mksquashfs tool since the initramfs are being build to include the target's DEFAULT_PACKAGES and not the individual DEVICE_PACKAGES.

musashino205 replied:
I see. Can I add ```ramdisk``` to ```FEATURES``` in the Makefile?

musashino205 replied:
I see, OK.

musashino205 replied:
I sent the patch to the mailing list.

chunkeey commented on target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -0,0 +1,512 @@
+		router {
+			label = "router";
+			gpios = <&qcom_pinmux 56 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_1>;
```
Q: This is one of those three-position switches, isn't it?

musashino205 replied:
Yes, it is one of the three-position switch.

chunkeey commented on target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh:
```diff
@@ -0,0 +1,182 @@
+create_fakeroot() {
+	local fkroot_dir="/tmp/fkroot"
+
+	if [ ! -x "/usr/sbin/mksquashfs" ]; then
```
Would it be possible to move this check into the preparation stage, before the ubi volumes are all removed?

chunkeey commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -72,6 +72,20 @@ define Device/ZyXELImage
+define Device/buffalo_wxr-2533dhp
+	$(call Device/LegacyImage)
+	DEVICE_DTS := qcom-ipq8064-wxr-2533dhp
+	BOARD_NAME := wxr-2533dhp
```
BOARD_NAME should only be necessary for new devices.

chunkeey commented on target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh:
```diff
@@ -0,0 +1,182 @@
+# needs "ubi_rootfs" as rootfs volume name. This name cannot use as
+# rootfs volume name in OpenWrt, so we need to prepare dummy rootfs
+# data with squashfs format. And, U-Boot checks the checksum at the
+# end of rootfs (ubi_rootfs).
```
maybe you can use the ubi_rootfs dummy as the main rootfs. OpenWrt does not strictly need the rootfs to be named "rootfs" (but if there is a rootfs it will do the automount). it can be anything you want (ubi volume name restriction apply) as long as you pass the kernel the right ubi.block=xyz via the [cmdline/bootargs](https://www.systutorials.com/linux-kernels/474146/ubi-r-o-block-driver-on-top-of-ubi-volumes-linux-3-15/).

chunkeey commented:
> I am waiting for a reaction on my last reply...

Why? I think the PR just needs a fix for the xorimage segfault and it would be ready for being merged.

musashino205 commented:
Changed to use ```ubi_rootfs``` volume as rootfs.

chunkeey commented:
> Changed to use ubi_rootfs volume as rootfs.

Looking at the buffalo_upgrade_prepare_ubi() and buffalo_do_upgrade() reminds me of the RT-AC58U. It too had an non-openwrt-standard sysupgrade at the time. But instead of writing custom
code, I pushed to make the [nand.sh sysupgrade code](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh) fit my needs with patches like 93aa8604052305830ea6801f0b93ffe2f0d55721 and  7783f31359cbd912245785e8861ceeb7feb9a8a1 (because the first one was incomplete).

Do you think you could maybe do the same and introduce a CI_ROOTPART which would allow you to reuse some of the functions?

musashino205 commented:
Actually, I also thought I could cope with *nand.sh* change.
However, in WXR-2533DHP, it is necessary to handle ```ubi_rootfs_data``` and ```kernel``` in second partition in addition to them, and I have not thought of how to handle with it yet...

chunkeey commented:
Isn't it possible to keep just the snippets from buffalo_upgrade_prepare_ubi() around that deal with ubi_rootfs_data and the kernel on the second ubi volume?



    buffalo_upgrade_prepare_ubi()
    {
    	local ubidev="$( nand_find_ubi ubi )"
    	local mtdnum2="$( find_mtd_index "rootfs_1" )"

    	if [ ! "$mtdnum2" ]; then
    		echo "cannot find second ubi mtd partition rootfs_1"
                return 1
    	fi

    	local ubidev2="$( nand_find_ubi rootfs_1 )"
    	if [ ! "$ubidev2" ] && [ -n "$mtdnum2" ]; then
    		ubiattach -m "$mtdnum2"
    		ubidev2="$( nand_find_ubi rootfs_1 )"
    	fi

    	ubirmvol /dev/$ubidev -N ubi_rootfs_data &> /dev/null || true
    	ubirmvol /dev/$ubidev2 -N kernel &> /dev/null || true
    }

and then do

     platform_do_upgrade() {
 	case "$(board_name)" in
 	buffalo,wxr-2533dhp)
                buffalo_upgrade_prepare_ubi
                CI_ROOTPART="ubi_rootfs"
 		nand_do_upgrade "$ARGV"
 		;;


(Of course, CI_ROOTPART would need to be added to OpenWrt's nand.sh too)?

musashino205 commented:
You are right, I seem to have been trapped in the idea of dealing *ubi_rootfs_data* and *kernel* in second partition within ```buffalo_upgrade_prepare_ubi``` (or ```nand_upgrade_prepare_ubi```).
I modified *buffalo.sh* and *nand.sh* as your suggestion, and it works correctly.

chunkeey commented on package/base-files/files/lib/upgrade/nand.sh:
```diff
@@ -10,6 +10,9 @@ CI_KERNPART="${CI_KERNPART:-kernel}"
 # 'ubi' partition on NAND contains UBI
 CI_UBIPART="${CI_UBIPART:-ubi}"
 
+# 'rootfs' partition on NAND contains the rootfs
```
Could you please separate out the CI_ROOTPART changes in nand.sh as a separate patch?

musashino205 replied:
OK, I'll do it.

chunkeey commented on target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh:
```diff
@@ -0,0 +1,45 @@
+
+buffalo_upgrade_prepare_ubi() {
+	local ubidev="$( nand_find_ubi ubi )"
+	local mtdnum2="$( find_mtd_index "rootfs_1" )"
```
Sorry, the double quotes around "rootfs_1" sneaked in because of my unlucky c&p'ing earlier. Since rootfs_1 is just as valid, you might as well remove the extra quotes. (But this is highly optional).

musashino205 committed:
base-files: allow non-standard rootfs volume name in UBI in sysupgrade

This commit allows to use non-standard UBI volume name as the rootfs
volume in sysupgrade.

ex.:
  The U-Boot on Buffalo WXR-2533DHP checks existence and checksum of
  "ubi_rootfs" volume when booting, so this name is required.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

chunkeey commented:
> Done.

I've added the patches to my staging repo:

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=80bcb8c458f8e4ebe8c5afd943f673b249e84c59 (ramdisk feature)

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=0dad18a403c534b5c551c7b4dfc0c17135271f07 (CI_ROOTPART)

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=be5a6a0f8de52c2bbd677e6e80fe2446b0ded6ee (device)

I've reworded some of the commit messages. Please take a look if
the information is still correct.

musashino205 commented:
Looks good to me, really thank you for your advices and works!

chunkeey commented:
Merged. Thanks.
